### PR TITLE
Add the Browser Terminal as an IDE

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description
Adds Xterm.js as a web alternative to connecting to your workspace via SSH from your terminal. The code can be found at https://github.com/gitpod-io/xterm-web-ide and the Docker image is at [ide/xterm-web](https://console.cloud.google.com/gcr/images/gitpod-core-dev/eu/build/ide/xterm-web).

> **Note**: The original PR is at https://github.com/gitpod-io/gitpod/pull/8464, but it was merged incorrectly. 

> **Note**: this relies on https://github.com/gitpod-io/gitpod/pull/17240 being merged first

Preview environment 🌐: https://ide-browser-terminal.preview.gitpod-dev.com/workspaces

## How to test
1. Open the preview environment
2. In the preferences, notice there is no Xterm IDE
3. In [Configcat](https://app.configcat.com/), include yourself in the enable-rule for `experimentalIdes` (non-prod)
4. Wait a couple of minutes (🫖 time)
5. Reload and see the Terminal IDE magically appear
6. Create a workspace and have a blast

## Useful commands

Re-pull the IDE: 
```sh
kubectl rollout restart deployment/ide-service
```

## Current limitations

- [x] Does not work on mobile [fixed]
	<details><summary>iPad demo</summary>
		
	https://user-images.githubusercontent.com/29888641/227277864-891126e7-de8a-4038-a801-9ea669167952.mp4
		

	</details> 

## Release Notes

```release-note
NONE
```
⬆️ we can mention it after we finish the internal dog food

## Todo

- [x] `gp stop` closes xterm.js editor and shows the "Workspace Stopping" screen 
- [x] `gp open` works
- [x] `gp preview` works
- [x] Ensure heartbreating works
- [x] Move Docker image from DockerHub to our GCP registry

### Nice-to-haves
- [ ] Local echo [[source](https://github.com/gitpod-io/openvscode-server/blob/gp-code/main/src/vs/workbench/contrib/terminal/browser/xterm/terminalTypeAheadAddon.ts)]

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/ide-service/pkg/server/ideconfig.go
Comment (akosyakov) [inline components/ide-service/pkg/server/ideconfig.go:56]: Should not it be per a user? 🤔 otherwise it either enables for all or none?

Comment (filiptronicek) [reply components/ide-service/pkg/server/ideconfig.go:56]: It should, do you know where to get the user data info from?

Comment (akosyakov) [reply components/ide-service/pkg/server/ideconfig.go:56]: It should be done in getConfig then, i.e. provide user id and email similar to resolveWorkspaceConfig.

**Review by gtsiolis — COMMENTED** [review]
Added my email on the feature flag but still not seeing the editor in `/new` or `/preferences`. 🫖

**Comment by gtsiolis** [comment]
Cross-posting for visibility a review comment from the original PR in https://github.com/gitpod-io/gitpod/pull/8464#pullrequestreview-1334749556:

> **thought**: Overall, UX LGTM and following the flow we're using with other editor options.

> **suggestion**: Also, since this option only supports one terminal, this seems more closely related to the connection with [Command Line](https://www.gitpod.io/docs/references/ides-and-editors/command-line) feature. Cc @loujaybee

> - This means users who would be interested in this will probably be interested in either using 🅰️ their local terminal or 🅱️ the browser xterm.
>     - This is also related to what a customer reported once in https://github.com/gitpod-io/gitpod/issues/7713#issuecomment-1140534710.
> - Maybe it's worth toning down this a bit and merging with the Command Line feature, strengthening the SSH functionality.
>   - This would make SSH a primary option as an editor and when the workspace is loaded the user could have the option to copy the access token or SSH key to connect with their local terminal or fallback to opening a terminal in the browser.
>   - We could also make it pop up as a smaller browser window which is the most common approach for such cases to emulate a terminal UX for users using the command line in browser.

**Comment by filiptronicek** [comment]
Hello @gtsiolis, could you please look at https://github.com/gitpod-io/gitpod/pull/17240 first? There are still a couple of things we need to figure out here after that one gets merged.

**Comment by filiptronicek** [comment]
@gtsiolis may I ask you to give this another go 🙏? I just tried and it seems to work 👀

File: components/supervisor/frontend/src/shared/loading-frame.ts
Comment (filiptronicek) [inline components/supervisor/frontend/src/shared/loading-frame.ts:19]: This solves a bug in which in Safari, the loading screen would not be hidden when the IDE loaded. It wasn't present in VS Code, because we actually append the IDE itself lazily, but for Xterm and potentially other web IDEs we might have a static HTML structure ready.

File: install/installer/pkg/components/ide-service/ide_config_configmap.go
Comment (filiptronicek) [inline install/installer/pkg/components/ide-service/ide_config_configmap.go:199]: Here we add the IDE to the IDE list. Currently, latest is the same as stable, since there are no proper versions yet.

File: components/ide-service/pkg/server/server.go
Comment (akosyakov) [inline components/ide-service/pkg/server/server.go:125]: We should not update like that. It is shared value for all users.

Comment (filiptronicek) [reply components/ide-service/pkg/server/server.go]: @akosyakov made changes in https://github.com/gitpod-io/gitpod/pull/17196/commits/12c1b0df83fff88d15abed4c2f8ae9560cb772ce, looks better?

Comment (akosyakov) [reply components/ide-service/pkg/server/server.go]: yes, not sure whether we need to preserve other properties of IDE config, both branches should return similar shapes to avoid surprises later that one return only one property but other all

Comment (akosyakov) [reply components/ide-service/pkg/server/server.go]: Why don't precompute 2 properties instead and then return either one or another?

Comment (filiptronicek) [reply components/ide-service/pkg/server/server.go]: Nice point for optimization! Let's do that

Comment (filiptronicek) [reply components/ide-service/pkg/server/server.go]: fc096e4d96419962c3def8b5de6f35920f39e9f3

Comment (akosyakov) [reply components/ide-service/pkg/server/server.go]: looks good, but you need to test :)

**Review by akosyakov — APPROVED** [review]
code-wise looks good, but I did not test

/hold

**Comment by filiptronicek** [comment]
Just tested the feature flag and it seems to work again
| No feature flag | Part of the feature flag |
|--------|--------|
| image | image |

**Review by easyCZ — APPROVED** [review]

**Pull request merged**
